### ISSUE TITLE
Add Real-Time Validation to Account Registration Form

### ISSUE THREAD

**Issue by mekarpeles** [internetarchive/openlibrary#2055]
Dependency to `epic` #7694 

# Considerations
UX side note: don't show negative feedback while a user is typing (don’t debounce during typing, instead show after e.g. tab out / defocus)

# Requirements
- [x] #9189 
  - PR: #9190  
- [x] #9165
  - PR: #9215
- [x] #7080
  - PR: #9223

### Is your feature request related to a problem? Please describe.

Add real-time, client-side validation for email, username, and password to the account registration form.

### Describe the solution you'd like

User should see feedback that:
- Email is unavailable on IA/OL 
- Username is already taken on IA/OL
- Email, username and/or password is incorrectly formatted

### Proposal & Constraints

Please try to use existing components as much as possible (minimize css changes)

**Comment by guyjeangilles** [comment]
Labeling this backlogged since it doesn't look like it's been worked on since May. Let me know your thoughts @mekarpeles

**Comment by rebecca-shoptaw** [comment]
👍🚀

**Comment by rebecca-shoptaw** [comment]
## Status update!
It's been great working on this project today, and I've got a good amount of progress to report so far.

For the sake of simplicity, I'll make and keep updating a checklist re: what we would like the form to do here:

### Required form behavior / tasks
- [ ] **If all fields are entered, we shouldn’t show an error that says “all fields required”** --> in progress in issue #9106 
- [ ] Error validation should happen before submit
  - [x] **Refactor/streamline of error rendering process** --> fixed by #9190 
  - [x] **If a badly formatted or already-taken email is entered, the error should appear before form submit**--> fixed by main project branch, pending further Docker tests
  - [x] **If a badly formatted or already-taken username is entered, the error should appear before form submit** --> fixed by main project branch, pending further Docker tests and investigation re: IA username validation
  - [ ] **If a badly formatted password is entered, the error should appear before form submit** --> to be fixed by #9165 
- [ ] Bad form submission should be stopped from the frontend side
  - [x] **Add `type="email"`, `required`, and regex `pattern` attributes to form inputs to prevent bad submission** --> fixed by main project branch by adding attributes to `forms.py` and a new form class to `form.py`
  - [ ] **Remove `validate` class from form if possible to avoid showing duplicate error messages as a result of HTML validation** --> pending approval
  - [ ] **Add `js` behavior to prevent submission (not just show an error msg) if backend checks like existing email address fail** --> in progress in main branch
  - [ ] **Remove remaining backend validation** --> pending review/approval
- [ ] Final overall cleanup of unused validation methods

### Other Requirements
- [x] **We need to communicate that screen name is public** --> already done, if a little inelegantly
- [x] **Clicking submit/pressing enter, any empty fields will be highlighted red, and the error text next to the field will say “This field is required”** --> fixed in main project branch via built-in HTML `required` attribute

### Fun possibilities
- [ ] Generate username suggestions using whatever function appends digits in the backend
- [ ] Add `keyup` checks to remove error message as soon as the input is fixed (but _not_ apply the error message on key up, see UX side-note in original issue)
- [ ] Disable submit button until all inputs are entered
- [ ] Do some kind of animation then check mark or x to indicate whether username is available (a very common and nice thing in forms)
